src/typemap.c

```diff
@@ -47,8 +47,9 @@ static int sig_match_by_type_simple(jl_value_t **types, size_t n, jl_tupletype_t
     for (i = 0; i < lensig; i++) {
         jl_value_t *decl = jl_field_type(sig, i);
         jl_value_t *a = types[i];
-        if (jl_is_type_type(decl)) {
-            jl_value_t *tp0 = jl_tparam0(decl);
+        jl_value_t *unw = jl_is_unionall(decl) ? ((jl_unionall_t*)decl)->body : decl;
```

Comment:
it's impossible for `decl` to be a unionall here (`jl_typemap_insert` doesn't know what this type is, so it won't classify it as a `simplesig`)